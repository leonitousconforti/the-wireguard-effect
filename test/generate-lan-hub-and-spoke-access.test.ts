/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { describe, expect, it } from "@effect/vitest";

import * as Effect from "effect/Effect";
import * as Schema from "effect/Schema";
import * as WireguardConfig from "the-wireguard-effect/WireguardConfig";
import * as GenerateExample from "../examples/generate-lan-hub-and-spoke-access.js";

describe("wireguard generate lan hub and spoke access", () => {
    it.effect("Should generate configs", () =>
        Effect.gen(function* () {
            const configs = yield* GenerateExample.program();
            expect(configs).toHaveLength(4);

            const config0 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[0]);
            const config1 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[1]);
            const config2 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[2]!);
            const config3 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[3]!);

            const keyMatcher = expect.stringMatching(/^[\d+/A-Za-z]{42}[048AEIMQUYcgkosw]=$/);
            const peerEntryMatcher = { PublicKey: keyMatcher, PresharedKey: keyMatcher };

            expect(config0).toMatchSnapshot({
                PrivateKey: keyMatcher,
                Peers: [peerEntryMatcher, peerEntryMatcher, peerEntryMatcher],
            });

            expect(config1).toMatchSnapshot({ PrivateKey: keyMatcher, Peers: [peerEntryMatcher] });
            expect(config2).toMatchSnapshot({ PrivateKey: keyMatcher, Peers: [peerEntryMatcher] });
            expect(config3).toMatchSnapshot({ PrivateKey: keyMatcher, Peers: [peerEntryMatcher] });
        })
    );
});
