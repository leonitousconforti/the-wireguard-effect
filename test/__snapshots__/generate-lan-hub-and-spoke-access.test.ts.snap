// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wireguard generate lan hub and spoke access > Should generate configs 1`] = `
{
  "Address": "10.0.0.1/24",
  "ListenPort": 51820,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.100/32",
      },
      "Endpoint": undefined,
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
    {
      "AllowedIPs": Set {
        "10.0.0.2/32",
      },
      "Endpoint": undefined,
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
    {
      "AllowedIPs": Set {
        "10.0.0.3/32",
      },
      "Endpoint": undefined,
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;

exports[`wireguard generate lan hub and spoke access > Should generate configs 2`] = `
{
  "Address": "10.0.0.100/24",
  "ListenPort": 0,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.1/32",
        "192.168.1.1/24",
      },
      "Endpoint": "server.wireguard.com:51820:51820",
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;

exports[`wireguard generate lan hub and spoke access > Should generate configs 3`] = `
{
  "Address": "10.0.0.2/24",
  "ListenPort": 0,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.1/32",
        "192.168.1.1/24",
      },
      "Endpoint": "server.wireguard.com:51820:51820",
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;

exports[`wireguard generate lan hub and spoke access > Should generate configs 4`] = `
{
  "Address": "10.0.0.3/24",
  "ListenPort": 0,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.1/32",
        "192.168.1.1/24",
      },
      "Endpoint": "server.wireguard.com:51820:51820",
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;
